Add relative linking support

mitchell-then opened the pull request:
- ham-fisted approach to relative linking
- if a link is a dox source file, change the link to a Confluence URL with page ID
- if a link is a file, change the link to point to repo_browse_url

mitchell-then commented on cmd/root.go:
```diff
@@ -44,7 +44,18 @@ to quickly create a Cobra application.`,
+			}
+		}
+
+		// ham-fisted approach to ensuring that all relative links are linked
```
I really don't like this, but without fundamental changes to the way dox does its publishing, I can't think of another workable solution right now.

jesselang replied:
Pros:
* Works
* ???

Cons:
* Publishes twice
* ???

Would you add anything to either of these lists? Is the primary reason for double publishing because there are no guarantees about the order of publishing, and potentially also recursive links (page A links to page B, page B links to page A)? If so, how would you go about improving this situation?

mitchell-then replied:
I'd say the primary reason is that we need to know the page IDs for the dox source files in order to relatively link to them. Same problem can be seen in [image support ](https://github.com/jesselang/dox/blob/master/pkg/publish.go#L92-L103) where when a page is first created, we can't attach files to it at the same time since the attachment endpoint needs a page ID.

This is a little different because a dox source file could link to another dox source file that hasn't been published yet, so the relative link would not be replaced. This way to guarantee that all dox source files have page IDs is to publish them all, then publish again for relative linking. I don't like it.

One alternative might be to perform some sort of pre-pass, where dox ensures that all dox source files are published (could be existing so no work, or new dox source so create a blank page) and in the main pass proceed to "update" all pages. This would also help with image support.

mitchell-then replied:
Latest commit has code to address this problem.

jesselang commented on pkg/publish.go:
```diff
@@ -250,3 +260,83 @@ func getBytesSha256(b []byte) string {
 	h.Write(b)
 	return hex.EncodeToString(h.Sum(nil))
 }
+
```
What do you think about moving all this to a `publish_links.go`? And maybe move the image functions to `publish_images.go`?

mitchell-then replied:
- [x] move linking and image code to separate files

jesselang commented on cmd/root.go:
```diff
@@ -34,7 +34,7 @@ to quickly create a Cobra application.`,
 		}
 
-		root, err := dox.Publish("", "", dryRun) // generated root page
+		root, err := dox.Publish("", "", repoRoot, dryRun) // generated root page
```
Where is `repoRoot` defined?

mitchell-then replied:
Declared here: https://github.com/jesselang/dox/blob/master/cmd/root.go#L15
Defined here: https://github.com/jesselang/dox/blob/master/cmd/root.go#L74

jesselang replied:
Of course! That was a silly question. Lol

jesselang commented:
Could you rebase your changes on master rather than merging master in?

jesselang commented on pkg/publish.go:
```diff
@@ -250,3 +260,83 @@ func getBytesSha256(b []byte) string {
+
+		src, err := source.New(localAnchorHrefPath, source.Opts{StripComments: true, TrimSpace: true})
+		if err != nil {
+			// file exists but is not a dox source file, link to stash instead
```
`s/stash/source/`?

mitchell-then replied:
- [x] s/stash/source/

jesselang commented on pkg/publish.go:
```diff
@@ -250,3 +260,83 @@ func getBytesSha256(b []byte) string {
+			// file exists but is not a dox source file, link to stash instead
+			localAnchorHrefFromRepoRoot := strings.Replace(localAnchorHrefPath, repoRoot, "", -1)
+			if strings.HasPrefix(localAnchorHrefFromRepoRoot, "/") {
+				localAnchorHrefFromRepoRoot = localAnchorHrefFromRepoRoot[1:]
```
Check out https://golang.org/pkg/strings/#TrimPrefix

mitchell-then replied:
- [x] Use `TrimPrefix`

jesselang commented on pkg/publish.go:
```diff
@@ -250,3 +260,83 @@ func getBytesSha256(b []byte) string {
+			pageContent = strings.Replace(pageContent, fmt.Sprintf(`href="%s"`, localAnchorHref), fmt.Sprintf(`href="%s/%s"`, repoBrowseURL, localAnchorHrefFromRepoRoot), -1)
+			continue
+		} else if src.ID() != "" {
+			pageContent = strings.Replace(pageContent, fmt.Sprintf(`href="%s"`, localAnchorHref), fmt.Sprintf(`href="%s/pages/viewpage.action?pageId=%s"`, uri, src.ID()), -1)
```
It doesn't feel right to have Confluence specific URL patterns embedded as a literal. I think it should be a const at least. I wouldn't mind making `dox` able to publish to Github Wiki or other targets in time, so the more we can keep from sprinkling target specific details all over, the easier it will be to abstract them into interfaces.

mitchell-then replied:
- [x] address Confluence URL

mitchell-then commented:
> 
> 
> Could you rebase your changes on master rather than merging master in?

- [x] rebase on master

mitchell-then committed:
Add relative linking support

jesselang commented on pkg/publish.go:
```diff
@@ -112,12 +109,17 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
+			return "", err
+		}
+
+		pageContent, err = ReplaceRelativeLinks(file, pageContent, uri, repoBrowseURL, repoRoot)
```
All these function calls can be private since they are within the same `dox` package.

mitchell-then replied:
This makes sense, I was wondering why I didn't have to import those files, but if its all same package then that helps me understand.

- [x] make functions private

jesselang commented on pkg/publish.go:
```diff
@@ -28,6 +20,11 @@ func Publish(file string, rootID string, dryRun bool) (id string, err error) {
 		return id, errors.New("space must be set in config")
 	}
 
+	repoBrowseURL := viper.GetString("repo_browse_url")
```
What do you think about `browse_url_base`? Or perhaps even better would be to allow users to specify a format string?

```
browse_uri: https://github.com/jesselang/dox/blame/master # which would append "/%s" if not present in the string
# which would allow extra parameters to be passed
browse_url: https://othersourcesite.com/repo-base/browse/%s?format=raw
```

mitchell-then replied:
I think that's a good idea, I'll see if I can do it.

- [x] change `repo_browse_url` `browse_url_base` support string formatting

mitchell-then committed:
Make package functions private

mitchell-then committed:
Allow formatting on URL for links to source files

mitchell-then committed:
Update documentation for relative linking

mitchell-then committed:
Split publishing into two passes

Split publishing into PrePublish and Publish to ensure that all pages exist to more efficiently publish page content.

jesselang commented on cmd/root.go:
```diff
@@ -34,28 +34,32 @@ to quickly create a Cobra application.`,
 		if err != nil {
 			fmt.Fprintf(os.Stderr, "error: %s\n", err)
 			os.Exit(1)
 		}
```
I don't think a `PrePublish` step is necessary for the root page.

mitchell-then replied:
We still need it in order to create the root page. Running dox on a "fresh" repo with no published pages would fail because it will try to update the root page which doesn't exist.

jesselang replied:
Gotcha.

jesselang reviewed (approved):
Thanks Mitch! Great job getting this feature added given the current process!
